File: insertion_sort/insertion_sort.go
package main

import "fmt"

// InsertionSort function returns sorted data
// Time complexity: O(n^2)
// Space complexity: O(1)
func InsertionSort(data []int) []int {
	for i := 1; i < len(data); i++ {
		key := data[i]
		j := i - 1
		for j >= 0 && data[j] > key {
			data[j+1] = data[j]
			j--
		}
		data[j+1] = key
	}
	return data
}

func main() {
	data := []int{1000, 45, -45, 121, 47, 45, 65, 121, -1, 103, 45, 34}
	data = InsertionSort(data)
	fmt.Println(data)
}

File: insertion_sort/insertionSort.js
/* Following algorithm sorts the input array in ascending order
* Time Complexity : O(n^2)
* Auxiliary Space: O(1)
* n is the number of elements in the array to be sorted
*/

function insertionSort (arr) {
	/*
	: param arr : Array to be sorted
	: return : Sorted Array
	*/
	for (let i = 1; i < arr.length; i++) {
		let j = i - 1;
		let temp = arr[i];
		while (j >= 0 && arr[j] > temp) {
			arr[j + 1] = arr[j];
			j--;
		}
		arr[j + 1] = temp;
	}
	return arr;
}

function main () {
	let arr = [5, 9, 3, 1, 99];
	arr = insertionSort(arr);
	console.log(arr);
}

main();
